File: arm.64/langs/ocaml/src/find.ml
let find target_dir =
  let dir =
    match target_dir with
    | Some d -> d
    | None -> (
        match Sys.getenv_opt "DIST_DIR" with
        | Some d -> d
        | None ->
            let candidates = ["../../dist"; "../../../dist"; "dist"] in
            let rec check = function
              | [] -> "dist"
              | c :: cs -> if Sys.file_exists c && Sys.is_directory c then c else check cs
            in check candidates
      )
  in
  if not (Sys.file_exists dir && Sys.is_directory dir) then []
  else
    Sys.readdir dir
    |> Array.to_list
    |> List.filter (fun f -> f <> ".DS_Store")
    |> List.map (Filename.concat dir)
    |> List.filter (fun f -> not (Sys.is_directory f))

File: arm.64/langs/haskell/src/Find.hs
module Find where

import System.Directory
import System.FilePath
import System.Environment
import Control.Monad

find :: Maybe String -> IO [FilePath]
find targetDir = do
  envDir <- lookupEnv "DIST_DIR"
  let candidates = ["dist", "../../dist", "../../../dist"]
  existing <- filterM doesDirectoryExist candidates
  let fallback = if null existing then "dist" else head existing
  let dir = case targetDir of
              Just d -> d
              Nothing -> maybe fallback id envDir
  exists <- doesDirectoryExist dir
  if not exists
    then return []
    else do
      entries <- listDirectory dir
      let valid = filter (/= ".DS_Store") entries
      filterM doesFileExist (map (dir </>) valid)
